File: 054-spiral-matrix/spiral-matrix.py
# -*- coding:utf-8 -*-


# Given a matrix of m x n elements (m rows, n columns), return all elements of the matrix in spiral order.
#
#
#
# For example,
# Given the following matrix:
#
#
# [
#  [ 1, 2, 3 ],
#  [ 4, 5, 6 ],
#  [ 7, 8, 9 ]
# ]
#
#
# You should return [1,2,3,6,9,8,7,4,5].


class Solution(object):
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        if not matrix:
            return []
            
        result = []
        while len(matrix)>0:
            t = matrix.pop(0)
            matrix = self.trans(matrix)
            result += t
        return result
        

    def trans(self, matrix):
        return list(zip(*matrix))[::-1]

File: 054-spiral-matrix/spiral-matrix.cpp
// Given a matrix of m x n elements (m rows, n columns), return all elements of the matrix in spiral order.
//
// Example 1:
//
//
// Input:
// [
//  [ 1, 2, 3 ],
//  [ 4, 5, 6 ],
//  [ 7, 8, 9 ]
// ]
// Output: [1,2,3,6,9,8,7,4,5]
//
//
// Example 2:
//
// Input:
// [
//   [1, 2, 3, 4],
//   [5, 6, 7, 8],
//   [9,10,11,12]
// ]
// Output: [1,2,3,4,8,12,11,10,9,5,6,7]
//


class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int re = matrix.size() - 1;
        if (re < 0) return {};
        int ce = matrix[0].size() - 1;
        if (ce < 0) return {};
        int rs = 0, cs = 0;
        vector<int> ans;
        while (rs <= re && cs <= ce) {
            for (int j = cs; j <= ce; j++) {
                ans.emplace_back(matrix[rs][j]);
            }
            rs++;
            for (int i = rs; i <= re; i++) {
                ans.emplace_back(matrix[i][ce]);
            }
            ce--;
            if (rs <= re){
            for (int j = ce; j >= cs; j--) {
                ans.emplace_back(matrix[re][j]);
            }
            re--;}
            if (cs <= ce) {
            for (int i = re; i >= rs; i--) {
                ans.emplace_back(matrix[i][cs]);
            }
            cs++;
            }
        }
        return ans;
    }
};
